File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
ListNode* merge(ListNode* a, ListNode* b) {
        ListNode* tempa = a;
        ListNode* tempb = b;
        ListNode* c = new ListNode(100);
        ListNode* tempc = c;
        while(tempa!=NULL && tempb!=NULL){
            if(tempa->val<=tempb->val){
                ListNode* t = new ListNode(tempa->val);
                tempc->next= t;
                tempc = t;
                tempa=tempa->next;

            }
            else{
               ListNode* t = new ListNode(tempb->val);
                tempc->next= t;
                tempc = t;
                tempb=tempb->next;  
            }

        }
        if(tempa==NULL){
            tempc->next=tempb;
        }
        else{
            tempc->next = tempa;
        }
        return c->next;
    }
    ListNode* mergeKLists(vector<ListNode*>& arr) {
        if(arr.size()==0) return NULL;
        while(arr.size()>1){
            ListNode* a =  arr[0];
            arr.erase(arr.begin());
            ListNode* b =  arr[0]; 
            arr.erase(arr.begin());
            ListNode* c = merge(a,b);
            arr.push_back(c);
        }
        return arr[0];
    }
};

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.java
// Definition for singly-linked list.
// public class ListNode {
//     int val;
//     ListNode next;
//     ListNode() {}
//     ListNode(int val) { this.val = val; }
//     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
// }

import java.util.*;

class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists == null || lists.length == 0) {
            return null;
        }
        
        // Min-heap based on node values
        PriorityQueue<ListNode> pq = new PriorityQueue<>(
            Comparator.comparingInt(n -> n.val)
        );
        
        // Add the head of each non-null list to the heap
        for (ListNode node : lists) {
            if (node != null) {
                pq.offer(node);
            }
        }
        
        ListNode dummy = new ListNode(0);
        ListNode tail = dummy;
        
        // Process nodes from heap
        while (!pq.isEmpty()) {
            ListNode min = pq.poll();
            tail.next = min;
            tail = min;
            
            // Add next node from the same list if exists
            if (min.next != null) {
                pq.offer(min.next);
            }
        }
        
        return dummy.next;
    }
}